FIX: Adaptive search bug that could trigger "Too many forward candidates"

nkeim opened the pull request:
This partially addresses #487. When many iterations of adaptive search are required to reduce a subnet to solvable size, there is a "leak" that adds superfluous null particles to the list of forward candidates. This PR fixes that bug, but performance in this case is still much worse than the adaptive search implementation in v0.3.

caspervdw commented:
This is looking great, I understand that now performance is even better than before?

apiszcz commented:
looking 3x faster compared to 0.3.3 and Friday ...

On Mon, Apr 9, 2018 at 2:04 AM, Casper van der Wel <notifications@github.com
> wrote:

> This is looking great, I understand that now performance is even better
> than before?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/soft-matter/trackpy/pull/489#issuecomment-379642737>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABXVTSretzOj30231qYQl0kD97kSlkM7ks5tmvnbgaJpZM4TLWDj>
> .
>

caspervdw commented on trackpy/linking/linking.py:
```diff
@@ -503,6 +500,9 @@ def next_level(self, coords, t, extra_data=None):
         for source_set, dest_set in self.subnets:
+            for sp in source_set:
+                sp.forward_cands.sort(key=lambda x: x[1])
+
```
Am I correct with assuming that not every subnetlinker requires the null particle? Else it's probably best done here?

caspervdw commented on trackpy/linking/subnet.py:
```diff
@@ -274,20 +280,19 @@ def split_subnet(source, dest, new_range):
         new_fcs = []
         for dp, dist in sp.forward_cands:
-            if dist <= new_range or dp is None:
+            if dist <= new_range:
```
This removes the null candidates, right? Maybe add comment to be clear about it

caspervdw commented on trackpy/linking/subnetlinker.py:
```diff
@@ -427,24 +427,41 @@ def subnet_linker_nonrecursive(source_set, dest_set, search_range, **kwargs):
-    # sort candidates and add in penalty for not linking
+    # Forward candidates were already sorted by Linker.assign_links()
     for _s in source_set:
-        _s.forward_cands.sort(key=lambda x: x[1])
```
The sort should be dropped in the other subnetlinkers as well

caspervdw commented on trackpy/linking/subnetlinker.py:
```diff
@@ -427,24 +427,41 @@ def subnet_linker_nonrecursive(source_set, dest_set, search_range, **kwargs):
+    # Forward candidates were already sorted by Linker.assign_links()
     for _s in source_set:
-        _s.forward_cands.sort(key=lambda x: x[1])
         _s.forward_cands.append((None, search_range))
```
Not sure if adding the null candidate should be done in the subnetlinker or before the subnetlinker. It seems cleaner to me to do it before? Are there any reasons for doing it like this?

caspervdw commented on trackpy/linking/utils.py:
```diff
@@ -100,6 +100,8 @@ class Point(object):
     example. )
     '''
+    __slots__ = ['_track', 'uuid', 't', 'pos', 'id', 'extra_data', 
+                 'forward_cands', 'subnet', 'relocate_neighbors', '__dict__']
```
Neat, I looked it up and I didn't know this feature

caspervdw commented on trackpy/tests/test_linking.py:
```diff
@@ -701,6 +701,12 @@ def setUp(self):
         self.linker_opts = dict(link_strategy='numba')
 
 
+class TestHybridLink(SubnetNeededTests):
```
Just a thought: what about making a mock subnetlinker and testing if the subnetlinker gets the expected subnets (sorted / not sorted, with / without null particle)?  In a separate PR of course

caspervdw commented on trackpy/linking/find_link.py:
```diff
@@ -475,6 +475,9 @@ def assign_links(self):
                 new_cands = set()
 
+            for sp in source_set:
+                sp.forward_cands.sort(key=lambda x: x[1])
```
Same comment as previous: maybe we want to add null particles over here.

caspervdw commented:
Fantastic!! I left some minor line notes

caspervdw commented:
@nkeim I found some time to write the test I commented about, also I made the null particle addition more consistent (and tested that)

nkeim commented:
@caspervdw Thanks! Sorry I have been absent this week. The mock subnet linker is a great idea. I just added a check that would have caught the original bug for this PR!

I added a reply to your comment about where in the code we should add the null candidate. Would it be OK if I reversed your commit? I think that the conceptual separation *and* slightly greater performance are worth the extra lines of code and duplication.

I'm sure you agree that this PR is getting messy! Once we are happy with it, I would like to rebase before merge.

caspervdw commented:
@nkeim sorry for causing the mess.. I am however not sure if I understand the performance difference... null particles are removed and added anyway by the adaptive step, that I did not change. It’s ok if you revert, maybe keep the commits and don’t rewrite history here ?

nkeim commented:
@caspervdw On the contrary, I started this mess!

It looks like your rollback commit did restore the code in `split_subnet()` that re-adds the null candidate. I have implemented the alternate scenario in which the (real) candidates are sorted in `linking.py`, and then the null candidate is added as an implementation detail in `subnetlinker.py`.

@apiszcz I think this will definitely be 0.4.1. I'm hoping we can release in the next few days.

caspervdw commented:
Ok, agreed! Let's look at #486 and then release v0.4.1
